File: internal/service/balance_log_service.go
package service

import (
	"CloudKey/internal/model"
	"time"

	"gorm.io/gorm"
)

type BalanceLogService struct {
	db *gorm.DB
}

func NewBalanceLogService(db *gorm.DB) *BalanceLogService {
	return &BalanceLogService{db: db}
}

type RecordBalanceParams struct {
	TenantID     uint64
	KeyID        uint64
	KeyAlias     string
	KeySuffix    string
	Delta        int64
	BeforeAmount int64
	AfterAmount  int64
	Operator     string
	Remark       string
}

func (s *BalanceLogService) Record(params RecordBalanceParams) error {
	return s.db.Create(&model.BalanceLog{
		TenantID:     params.TenantID,
		KeyID:        params.KeyID,
		KeyAlias:     params.KeyAlias,
		KeySuffix:    params.KeySuffix,
		Delta:        params.Delta,
		BeforeAmount: params.BeforeAmount,
		AfterAmount:  params.AfterAmount,
		Operator:     params.Operator,
		Remark:       params.Remark,
		CreatedAt:    time.Now(),
	}).Error
}

type BalanceLogQuery struct {
	Page      int    `form:"page"`
	PageSize  int    `form:"page_size"`
	Alias     string `form:"alias"`      // 别名前缀搜索
	KeySuffix string `form:"key_suffix"` // 后缀精准搜索
	StartTime string `form:"start_time"`
	EndTime   string `form:"end_time"`
}

func (s *BalanceLogService) ListLogs(query BalanceLogQuery, tenantID uint64) ([]model.BalanceLog, int64, error) {
	var logs []model.BalanceLog
	var total int64

	db := s.db.Model(&model.BalanceLog{}).Where("tenant_id = ?", tenantID)
	if query.Alias != "" {
		db = db.Where("key_alias LIKE ?", query.Alias+"%")
	}
	if query.KeySuffix != "" {
		db = db.Where("key_suffix = ?", query.KeySuffix)
	}
	if query.StartTime != "" {
		db = db.Where("created_at >= ?", query.StartTime)
	}
	if query.EndTime != "" {
		db = db.Where("created_at <= ?", query.EndTime)
	}

	if err := db.Count(&total).Error; err != nil {
		return nil, 0, err
	}

	offset := (query.Page - 1) * query.PageSize
	if err := db.Order("created_at DESC").Offset(offset).Limit(query.PageSize).Find(&logs).Error; err != nil {
		return nil, 0, err
	}
	return logs, total, nil
}

func (s *BalanceLogService) ExportLogs(query BalanceLogQuery, tenantID uint64) ([]model.BalanceLog, error) {
	var logs []model.BalanceLog
	db := s.db.Model(&model.BalanceLog{}).Where("tenant_id = ?", tenantID)
	if query.Alias != "" {
		db = db.Where("key_alias LIKE ?", query.Alias+"%")
	}
	if query.KeySuffix != "" {
		db = db.Where("key_suffix = ?", query.KeySuffix)
	}
	if query.StartTime != "" {
		db = db.Where("created_at >= ?", query.StartTime)
	}
	if query.EndTime != "" {
		db = db.Where("created_at <= ?", query.EndTime)
	}
	if err := db.Order("created_at DESC").Find(&logs).Error; err != nil {
		return nil, err
	}
	return logs, nil
}

File: sdk/python/cloudkey/balance_log_service.py
"""余额日志服务。"""

from __future__ import annotations

from typing import Any

from ._client import _HttpClient
from ._models import BalanceLog, PageResult


class BalanceLogService:
    """余额日志服务，提供余额变更日志的查询和导出。

    示例::

        logs = ck.balance_logs.list(page=1, page_size=20, alias="my-key")
    """

    def __init__(self, client: _HttpClient) -> None:
        self._client = client

    def list(
        self,
        *,
        page: int = 1,
        page_size: int = 20,
        alias: str | None = None,
        key_suffix: str | None = None,
        start_time: str | None = None,
        end_time: str | None = None,
    ) -> PageResult[BalanceLog]:
        """分页查询余额变更日志。

        Args:
            page: 页码，默认 1。
            page_size: 每页数量，默认 20。
            alias: 别名前缀搜索。
            key_suffix: 后缀精准搜索。
            start_time: 开始时间，格式 "2006-01-02 15:04:05"。
            end_time: 结束时间，格式 "2006-01-02 15:04:05"。

        Returns:
            分页结果。
        """
        params: dict[str, Any] = {"page": page, "page_size": page_size}
        if alias is not None:
            params["alias"] = alias
        if key_suffix is not None:
            params["key_suffix"] = key_suffix
        if start_time is not None:
            params["start_time"] = start_time
        if end_time is not None:
            params["end_time"] = end_time

        data = self._client.get("/service/balance-logs", params=params)
        d = data["data"]
        items = [_to_balance_log(l) for l in (d.get("list") or [])]
        return PageResult(
            items=items,
            total=d.get("total", 0),
            page=d.get("page", 1),
            page_size=d.get("page_size", 20),
        )

    def export(
        self,
        *,
        alias: str | None = None,
        key_suffix: str | None = None,
        start_time: str | None = None,
        end_time: str | None = None,
    ) -> list[dict[str, Any]]:
        """导出余额变更日志。

        Args:
            alias: 别名前缀搜索。
            key_suffix: 后缀精准搜索。
            start_time: 开始时间。
            end_time: 结束时间。

        Returns:
            日志列表。
        """
        params: dict[str, Any] = {}
        if alias is not None:
            params["alias"] = alias
        if key_suffix is not None:
            params["key_suffix"] = key_suffix
        if start_time is not None:
            params["start_time"] = start_time
        if end_time is not None:
            params["end_time"] = end_time

        data = self._client.get("/service/balance-logs/export", params=params)
        return data.get("data", []) or []


# ==================== 内部工具 ====================


def _to_balance_log(d: dict[str, Any]) -> BalanceLog:
    return BalanceLog(
        id=d.get("id", 0),
        key_id=d.get("key_id", 0),
        key_alias=d.get("key_alias", ""),
        key_suffix=d.get("key_suffix", ""),
        delta=d.get("delta", 0),
        before_amount=d.get("before_amount", 0),
        after_amount=d.get("after_amount", 0),
        operator=d.get("operator", ""),
        remark=d.get("remark", ""),
        created_at=d.get("created_at"),
    )
